### PR TITLE
Add dev guideline limiting auto usage.

### PR DESCRIPTION
**Pull request by TheBlueMatt**
I'm tired of having an extra level of indirection to review stupid shit like https://github.com/bitcoin/bitcoin/pull/11403/commits/5ae4080cc2c0c50e1093b52005c47cd74bd0261e#diff-ad6efdc354b57bd1fa29fc3abb6e2872R283, it just makes review, scripted-diff and other stuff harder, without materially making the code more flexible (unless we want to do a massive sed to make everything auto, which would be insane).

### CODE REVIEW

**Review by jonasschnelli — APPROVED** [review]
ACK 9398b3d2dbaa35f1f029d4072dad9282920430c8

File: doc/developer-notes.md [hunk 1]
```diff
@@ -39,6 +39,7 @@ code.
   - `++i` is preferred over `i++`.
   - `nullptr` is preferred over `NULL` or `(void*)0`.
   - `static_assert` is preferred over `assert` where possible. Generally; compile-time checking is preferred over run-time checking.
+  - Do not use `auto` unless it saves significant typing/reading or in method signatures where it makes the method more flexible.
```
Comment (promag) [inline doc/developer-notes.md]: µnit ``Only use `auto` if it ...``

**Review by promag — COMMENTED** [review]
ACK.

**Comment by sipa** [comment]
"auto" in method signatures in C++11 is only allowed when still specifying the return type explicitly (auto fn(args...) -> returntype notation). Is that what you're referring to?

**Comment by TheBlueMatt** [comment]
Ok, amended to say "or if it is critical to the functionality of the code", nice big loophole but if it really matters somewhere of course people can use it.

**Comment by MarcoFalke** [comment]
ACK 6e1fa4e50d99803b7150891cc53761191303bacf

**Comment by ajtowns** [comment]
This seems like it would be better if it was a bit more explicit/objective about what's acceptable:

> Use of `auto` outside of unit tests should be limited to the following circumstances in order to avoid hiding type information:
>      * range-based for loops, eg `for (const auto& i : vFoo)`
>      * when dealing with iterators in general, eg `auto it = vFoo.rbegin()`
>      * when creating a closure/lambda
>      * in templates, where necessary
>      * when using a variable as read-only alias of a long expression, eg `const auto x = long(expression())`

Presumably overusing auto in unit tests isn't a big deal and the first four bullet points are uncontroversial? 

Presumably the last one is what irritates @TheBlueMatt, so shouldn't be there, and places in the code that currently use it should get converted to specifying an actual type.

FWIW, in current master, usage of auto seems to pretty much match up with the list above so not having loopholes seems practical:

* A bunch of uses in C++11 for loops (133):
     - for (const auto& foo : bar) -- 7 in *.h, 94 in *.cpp
     - for (auto&& foo : bar) and for (auto& foo:bar)-- 4 in net.h, 1 in undo.h, 16 in *.cpp
     - for (auto foo : bar) -- 1 in reverse_iterator.h, 9 in *.cpp
     - for (const auto foo : bar) -- 1 in rpc/blockchain.cpp
* A bunch of uses with iterators (72):
     - auto it = foo.find(bar) -- 47 in *.cpp
     - auto it = foo.begin() (or rbegin) -- 11 in *.cpp
     - for (auto x = foo.begin(); ..) (or rbegin) -- 6 in *.cpp
     - another 8 that are probably iterator related: 
```
./support/lockedpool.cpp:    auto it = std::find_if(chunks_free.begin(), chunks_free.end(),
./support/lockedpool.cpp:    auto next = chunks_free.upper_bound(freed.first);
./support/lockedpool.cpp:    auto alloced = chunks_used.emplace(it->first + it->second - size, size).first;
./txmempool.cpp:        auto iter = mapNextTx.lower_bound(COutPoint(hash, 0));
./txmempool.cpp:        auto iter = mapNextTx.lower_bound(COutPoint(it->GetTx().GetHash(), 0));
./txmempool.cpp:    auto iters = GetSortedDepthAndScore();
./txmempool.cpp:    auto iters = GetSortedDepthAndScore();
./wallet/wallet.cpp:    auto iter = mapTxSpends.lower_bound(COutPoint(txid, 0));
```
* Shortcuts in bench/ and test/ (48)
     - for (auto x = 0; ...) -- 12 of them.. weird
     - 36 others
* Aliasing (22)
     - const auto foo = bar; -- 22 in *.cpp
* Usage with closures/lambdas (4):
     - auto foo = \[...\](...) { ... } -- 4 in *.cpp
* Usage with templates (3):
     - template return values -- 3 in reverse_iterator.h
* And 14 others in non-bench non-test *.cpp:
```
./base58.cpp:    auto bech = bech32::Decode(str);     // pair<string,vector<uint8>>
./httpserver.cpp:    auto req_copy = req;    // unique_ptr<HTTPRequest>
./net_processing.cpp:    auto ret = mapOrphanTransactions.emplace(hash, COrphanTx{tx, peer, GetTime() + ORPHAN_TX_EXPIRE_TIME});  // pair<iterator,bool>
./net_processing.cpp:                auto txinfo = mempool.info(inv.hash);  // TxMempoolInfo
./net_processing.cpp:                auto vtxinfo = mempool.infoAll();  // vector<TxMempoolInfo>
./net_processing.cpp:                    auto txinfo = mempool.info(hash);  // TxMempoolInfo
./net_processing.cpp:                        auto ret = mapRelay.insert(std::make_pair(hash, std::move(txinfo.tx))); // pair<iterator,bool>
./qt/optionsmodel.cpp:            auto ip_port = GetProxySetting(settings, "addrProxy"); // ProxySetting
./qt/optionsmodel.cpp:            auto ip_port = GetProxySetting(settings, "addrProxy");
./qt/optionsmodel.cpp:            auto ip_port = GetProxySetting(settings, "addrSeparateProxyTor");
./qt/optionsmodel.cpp:            auto ip_port = GetProxySetting(settings, "addrSeparateProxyTor");
./qt/paymentserver.cpp:                auto tempChainParams = CreateChainParams(CBaseChainParams::MAIN); // unique_ptr<CChainParams>
./qt/walletmodel.cpp:        auto& resultGroup = mapCoins[QString::fromStdString(EncodeDestination(group.first))]; // COutput
./support/lockedpool.cpp:    auto freed = *i; // pair<char* const, size_t>
```

**Comment by TheBlueMatt** [comment]
* Honestly I'm really not a big fan of use in for loops either.
* I don't think "iterators in general" is a reasonable metric, the metric is still readability-based - if the iterator is short (which they rarely are) it could still be more readable to type it out, if the iterator is a mile long (which they usually are) its obviously less readable.
* Same for a lambda.
* I'd strongly prefer if *nothing* used auto in return types, not just in templates. It makes review much, much harder.
* Again, its all about readability. Just because an expression is long does not mean it should use auto, the example I cited was to use auto for this precise reason but I found it quite annoying.

As for unit tests, I dont see why unit tests should be an exception, though we often simply dont apply style guidelines on unit tests anyway just because they have a lower bar.

**Comment by ajtowns** [comment]
The main difference I see between "auto x = FunctionCall(foo);" and "auto x = foo.begin()" or "for (auto x : foo)" is you can get a good expectation of the inferred type locally -- you just have to look back to the recent declaration of foo to know what convention implies the type of x will be; while with the function call you have to search header files for FunctionCall's declaration.

Lambdas/closures are similar in that their definition is right there, with the added problem that they have anonymous types in the first place, and apparently converting them to something you can write out has a performance hit.

I don't see what's ambiguous or confusing about "auto begin() const -> decltype(this->m_x.rbegin())", or how it's hard to review, or how you'd rewrite it to be better? C++14 type inference from the returned value would be worse though.

I do agree with your rationale here (more so after reviewing how auto's actually used), but "readability" is very subjective -- if I weren't following this discussion, I'd probably think "const auto& x = Blah();" "saves significant reading" compared to "const std::map<std::string,std::vector<std::string>>& x = Blah();" and end up violating the style guidelines while trying to comply with them. Getting called out for it during review would then be kind of irritating.

For the record, I had a go at a patch to switch unnecessary uses of auto to explicit types, at https://github.com/TheBlueMatt/bitcoin/pull/9

**Comment by theuni** [comment]
I generally disagree with discouraging 'auto' unless it completely obscures away the type.

See #12169 for a perfect example of the types of mistakes it can obsolete.

**Comment by laanwj** [comment]
> I generally disagree with discouraging 'auto' unless it completely obscures away the type.

I agree here. `auto` was a great addition, it both saves typing and verbosity while reading code. Yes, sometimes auto makes reviewing harder, in those specific cases do make a review comment, but right now this is a blanket discouragement that doesn't give any specific advice or rationale. So tend to NACK.

**Comment by TheBlueMatt** [comment]
>  it both saves typing and verbosity while reading code.
> this is a blanket discouragement that doesn't give any specific advice or rationale

Huh? This isn't a "blanket discouragement", it specifically calls out places where readability is better as places where auto is fine. I'm happy to reword it if you think it reads as a blanket discouragement, but that isn't the point.

On the other hand, I find the "readability" argument suspect except for cases like iterators and anonymous functions, though the "no implicit conversion" feature is quite nice, I'm less-than-convinced its worth trading being explicit to get it.

**Comment by laanwj** [comment]
> This isn't a "blanket discouragement"

To me it reads like a blanket discouragement, because you are just adding a line and not quoting any specific cases. It amounts to "don't use language feature X unnecessarily" which should be unnecessary to state.

**Comment by MarcoFalke** [comment]
Closing for now due to inactivity

**Pull request closed without merging**